runimagej.py

```diff
@@ -0,0 +1,359 @@
+            value = self._output_value(setting, result[details["name"]], self._ij)
+
+            # Record output images if they should be shown.
+            if isinstance(setting, cellprofiler.setting.ImageNameProvider):
+                image = cellprofiler.image.Image(value)
+                workspace.image_set.add(setting.value, image)
+                if self.show_window:
+                    workspace.display_data.images.append(value)
+
+    def settings(self):
+        settings = [
+            self.host,
+            self.connect,
+            self.input_count,
+            self.ij_module,
+            self.divider
+        ]
+
+        settings += self.input_settings.settings
+        settings += self.output_settings.settings
+
+        return settings
+
+    def validate_module(self, pipeline):
+        if not self._ij:
+            raise cellprofiler.setting.ValidationError(
+                "Not connected to host: {:s}\n\n"
+                "Please ensure the address is correct and the"
+                " ImageJ server is running.".format(self.host.value),
+                self.host
+            )
+
+    def visible_settings(self):
+        visible_settings = [
+            self.host,
+            self.connect
+        ]
+
+        if self._ij:
+            visible_settings += [self.ij_module]
+            visible_settings += self.input_settings.settings
+
+            if len(self.output_settings.settings) > 0:
+                visible_settings.append(self.divider)
+                visible_settings += self.output_settings.settings
+
+        return visible_settings
+
+    def _clamp(self, value, minval):
+        if value:
+            return value
+
+        return minval if minval else 0
+
+    def _connect(self, ij_module=None):
+        try:
+            self._ij = imagej.IJ(self.host.value)
+        except RuntimeError:
+            self._ij = None
+            return
+
+        self.ij_module = cellprofiler.setting.Choice(
+            "ImageJ module",
+            choices=self._get_ij_modules()
+        )
+
+        if not ij_module:
+            ij_module = self.ij_module.value
+
+        self._create_ij_settings(ij_module)
+
+    def _create_ij_settings(self, module_name):
+        self.input_details = []
+        self.input_settings = cellprofiler.setting.SettingsGroup()
+
+        self.output_details = []
+        self.output_settings = cellprofiler.setting.SettingsGroup()
+
+        module = self._ij.find(module_name)[0]
+        details = self._ij.detail(module)
+
+        # FOR DEBUGGING
+        logger.debug(json.dumps(details, indent=4))
+
+        for input_ in details["inputs"]:
+            name = input_["name"]
+            input_["rawType"] = raw_type = input_["genericType"].split("<")[0].split(" ")[-1]
+
+            # HACK: For now, we skip service and context parameters.
+            # Later, the ImageJ Server will filter these out for us.
```

Comment:
If there's a GitHub issue for ImageJ or something to link here, that would be excellent!

Reply:
Indeed—the issue to follow is imagej/imagej-server#14.